### PR TITLE
kvserver: eagerly extend expiration leases in Raft scheduler

### LINKED ISSUE #99812: kvserver: expired expiration leases prevent closed timestamp updates
Expired expiration leases can prevent closed timestamp updates:

https://github.com/cockroachdb/cockroach/blob/60b7440ec5b5d4647c53d180152e625e65e24a74/pkg/kv/kvserver/replica_closedts.go#L61-L67

This can be problematic when enabling `kv.expiration_leases_only.enabled`, because we don't eagerly extend leases unless there's traffic on the range, preventing the closed timestamp from advancing. The lease scheduler in #98433 would mitigate this.

Jira issue: CRDB-26097

Epic CRDB-25200

### PR DESCRIPTION
**Pull request by erikgrinaker**
This is submitted with a 23.1 backport in mind, and is therefore limited in scope. More work is needed to fully address #98433 in the 23.2 timeframe.

---

**kvserver: don't quiesce ranges with expiration-based leases**

This patch prevents quiescence with expiration-based leases, since we'll have to propose a lease extension shortly anyway. Out of caution, we only do this when `kv.expiration_leases_only.enabled` is `true`, with an eye towards 23.1 backports. For 23.2, we will always do this.

Touches #94592.

Epic: none
Release note: None
  
**kvserver: eagerly extend expiration leases in Raft scheduler**

This patch eagerly extends expiration leases in the Raft scheduler, during Raft ticks, but only if `kv.expiration_leases_only.enabled` is `true`. This is possible because we no longer allow ranges with expiration leases to quiesce in this case.

The old store lease renewer is disabled in this case, but still tracks local system ranges with expiration leases in case the setting is later changed. The store lease renewer is still retained for use by default, to allow backporting this change to 23.1 while keeping the existing default behavior.

Doing this during Raft scheduling is compelling, since it already does efficient goroutine scheduling (including liveness range prioritization) and holds the replica mutex.

Compared to an alternative implementation that relied on a separate, improved store lease renewer which also allowed ranges to quiesce between extensions, this has significantly better performance. The main reason for this is that unquiescing involves a Raft append to wake the leader. On an idle cluster with 20.000 ranges using only expiration leases, CPU usage was 26% vs.  30%, write IOPS was 500 vs. 900, and network traffic was 0.8 MB/s vs.  1.0 MB/s. KV benchmark results:

```
name                    old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3    14.6k ± 3%   15.2k ± 4%  +3.72%  (p=0.003 n=8+8)
kv95/enc=false/nodes=3   30.6k ± 3%   32.2k ± 3%  +5.22%  (p=0.000 n=8+8)

name                    old p50      new p50      delta
kv0/enc=false/nodes=3     11.3 ± 3%    11.0 ± 0%  -2.76%  (p=0.006 n=8+6)
kv95/enc=false/nodes=3    4.59 ± 8%    4.55 ± 3%    ~     (p=0.315 n=8+8)

name                    old p95      new p95      delta
kv0/enc=false/nodes=3     31.5 ± 0%    30.2 ± 4%  -4.25%  (p=0.003 n=6+8)
kv95/enc=false/nodes=3    19.5 ± 7%    18.1 ± 4%  -7.28%  (p=0.000 n=8+7)

name                    old p99      new p99      delta
kv0/enc=false/nodes=3     48.7 ± 3%    47.4 ± 6%    ~     (p=0.059 n=8+8)
kv95/enc=false/nodes=3    32.8 ± 9%    30.2 ± 4%  -8.04%  (p=0.001 n=8+8)
```

Resolves #99812.
Touches #98433.

Epic: none
Release note: None
  
**kvserver: also acquire expiration leases in replicate queue**

Previously, the replicate queue would only reacquire epoch leases, and allow expiration leases to expire. It now reacquires all leases, since they are eagerly extended.

In the future, when quiescence is removed entirely, this responsibility should be moved to the Raft tick loop.

Epic: none
Release note: None
  
**kvserver: conditionally extend expiration leases during requests**

If `kv.expiration_leases_only.enabled` is `true`, this patch disables expiration lease extension during request processing, since the Raft scheduler will eagerly extend leases in this case. This will eventually be removed completely, but is kept for now with an eye towards a 23.1 backport.

Epic: none
Release note: None

### CODE REVIEW

**Comment by erikgrinaker** [comment]
@andrewbaptist @nvanbenschoten As we discussed, I've updated the PR to gate all of the new behavior here on `kv.expiration_leases_only.enabled`. By default it retains practically all existing behavior.

**Review by andrewbaptist — COMMENTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

**Review by andrewbaptist — COMMENTED** [review]
Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-:-NSBZQSh472nEWtjHF1b:b-ucfknh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

**Review by andrewbaptist — APPROVED** [review]
:lgtm: This will be great to have in place!

> Reviewed 2 of 2 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-:-NSBeYuZ9YVEaLdOOyvH:b9iq3m5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_raft.go` line 1200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-NSBaJ7pCAe1EeL2cy2n:-NSBaJ7pCAe1EeL2cy2o:b-tajro6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c4f1f8d0f3ed872a3b20668bac19e8254b22fa/pkg/kv/kvserver/replica_raft.go#L1200)):*
> ```Go
> 	}
> 
> 	if r.mu.quiescent {
> ```

I'm unsure if we would ever get here with an expiration lease, but if we did get here and it is an expiration lease (possibly upgrade path?) should we handle this case by creating/extending a lease? I'm thinking of cases like upgrades or setting changes. 
___
*[`pkg/kv/kvserver/replica_range_lease.go` line 1417 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-NSBapnG8YEKBXl6TY7Q:-NSBapnG8YEKBXl6TY7R:b-rk79p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c4f1f8d0f3ed872a3b20668bac19e8254b22fa/pkg/kv/kvserver/replica_range_lease.go#L1417)):*
> ```Go
> // shouldExtendLeaseRLocked determines whether the lease should be extended
> // asynchronously, even if it is currently valid. The method returns true if
> // this range uses expiration-based leases, it is or was owned by this replica,
> ```

nit: This sentence is hard to parse now. Perhaps break it into conditions like:
The method returns true if the following conditions are all met:
* This range uses expiration-based leases.
* The most recent lease is owned by this replica.
* The lease is in need of renewal.
* There is no pending extension.
___
*[`pkg/kv/kvserver/replica_lease_renewal_test.go` line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-NSBcETQ9FRX6e1u09sx:-NSBcETQ9FRX6e1u09sy:bvbthgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c4f1f8d0f3ed872a3b20668bac19e8254b22fa/pkg/kv/kvserver/replica_lease_renewal_test.go#L55)):*
> ```Go
> 				RaftConfig: base.RaftConfig{
> 					RangeLeaseDuration: time.Second,
> 					RaftTickInterval:   100 * time.Millisecond,
> ```

nit: No need to change anything here, but it would be nice if we could "manually" call tick on all ranges. Something similar to `store.Enqueue`. It could make these tests much more deterministic and faster and could use a `ManualClock` instead of having to deal with all these `Eventually` calls below.

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100430)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_raft.go` line 1200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-NSBaJ7pCAe1EeL2cy2n:-NSCb7eQ3JXNY0Xq64y1:bnl6ial) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c4f1f8d0f3ed872a3b20668bac19e8254b22fa/pkg/kv/kvserver/replica_raft.go#L1200)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

I'm unsure if we would ever get here with an expiration lease, but if we did get here and it is an expiration lease (possibly upgrade path?) should we handle this case by creating/extending a lease? I'm thinking of cases like upgrades or setting changes. 
</blockquote></details>

If the setting is changed then the replicate queue will convert the lease to an expiration lease, which will wake up the Raft group, after which it will no longer quiesce. Beyond that, we don't currently attempt to eagerly acquire leases until the next client request arrives (or if the replicate queue gets to it within 10 minutes). This is equivalent to how we don't eagerly acquire epoch leases if they lapse either. We keep the scope of the Raft scheduler deliberately conservative and narrow for 23.1, so that it only extends leases once they're acquired, but we should improve this for 23.2 to make sure we always eagerly acquire leases on all ranges within seconds.
___
*[`pkg/kv/kvserver/replica_range_lease.go` line 1417 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-NSBapnG8YEKBXl6TY7Q:-NSCfFPB18UG05MY6DDv:b-dtkdrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c4f1f8d0f3ed872a3b20668bac19e8254b22fa/pkg/kv/kvserver/replica_range_lease.go#L1417)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: This sentence is hard to parse now. Perhaps break it into conditions like:
The method returns true if the following conditions are all met:
* This range uses expiration-based leases.
* The most recent lease is owned by this replica.
* The lease is in need of renewal.
* There is no pending extension.

</blockquote></details>

Done, thanks.
___
*[`pkg/kv/kvserver/replica_lease_renewal_test.go` line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/100430#-NSBcETQ9FRX6e1u09sx:-NSCfVOzDREpvqLU5vFG:b4oa1yy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c4f1f8d0f3ed872a3b20668bac19e8254b22fa/pkg/kv/kvserver/replica_lease_renewal_test.go#L55)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: No need to change anything here, but it would be nice if we could "manually" call tick on all ranges. Something similar to `store.Enqueue`. It could make these tests much more deterministic and faster and could use a `ManualClock` instead of having to deal with all these `Eventually` calls below. 
</blockquote></details>

Yeah, a lot of this code could be more amenable to unit-testing -- it would be nice to more easily test replica ticking and stepping in isolation, e.g. via a data-driven test or something. Although I also think there's value in having an integration test that wires everything up and makes sure it actually works. Either way, not a fight I'm going to pick now.
